server/src/main/java/org/elasticsearch/repositories/RepositoryData.java

```diff
@@ -522,150 +529,229 @@ public IndexMetaDataGenerations indexMetaDataGenerations() {
-                                        }
-                                    } else {
-                                        // the new format post 5.4.1 that only contains the snapshot uuid,
-                                        // since we already have the name/uuid combo in the snapshots array
-                                        uuid = parser.text();
-                                    }
-
-                                    SnapshotId snapshotId = snapshots.get(uuid);
-                                    if (snapshotId != null) {
-                                        snapshotIds.add(snapshotId);
-                                    } else {
-                                        // A snapshotted index references a snapshot which does not exist in
-                                        // the list of snapshots. This can happen when multiple clusters in
-                                        // different versions create or delete snapshot in the same repository.
-                                        throw new ElasticsearchParseException("Detected a corrupted repository, index " + indexId
-                                            + " references an unknown snapshot uuid [" + uuid + "]");
-                                    }
-                                }
-                            } else if (SHARD_GENERATIONS.equals(indexMetaFieldName)) {
-                                XContentParserUtils.ensureExpectedToken(
-                                    XContentParser.Token.START_ARRAY, parser.currentToken(), parser::getTokenLocation);
-                                while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
-                                    gens.add(parser.textOrNull());
-                                }
-
-                            }
-                        }
-                        assert indexId != null;
-                        indexSnapshots.put(indexId, Collections.unmodifiableList(snapshotIds));
-                        for (int i = 0; i < gens.size(); i++) {
-                            String parsedGen = gens.get(i);
-                            if (fixBrokenShardGens) {
-                                parsedGen = ShardGenerations.fixShardGeneration(parsedGen);
-                            }
-                            if (parsedGen != null) {
-                                shardGenerations.put(indexId, i, parsedGen);
-                            }
-                        }
-                    }
-                } else if (INDEX_METADATA_IDENTIFIERS.equals(field)) {
+        Map<String, String> indexMetaIdentifiers = null;
+        while (parser.nextToken() == XContentParser.Token.FIELD_NAME) {
+            final String field = parser.currentName();
+            switch (field) {
+                case SNAPSHOTS:
+                    parseSnapshots(parser, snapshots, snapshotStates, snapshotVersions, indexMetaLookup);
+                    break;
```

Comment:
Can we assert the state we are in after calling the parsing methods?

Reply:
It's tricky IMO. We can't really assert this I think (because corrupted `RepositoryData` would trip the assertion) so it's not really correct to assert.
My thinking was that https://github.com/elastic/elasticsearch/pull/62684/files#diff-98674f75d80495aed41485d9103543e0R712 eventually ensures that the parser parsed the `snapshots` field first and figured that's enough and we don't need another runtime check for it?

Reply:
OK, let's keep it that way then.